### PR TITLE
Restructure CHILLED code to be more trackable and customizable 

### PR DESCRIPTION
**Pull request by measrainsey**
Reorganize CHILLED code to be more package-friendly 

Main changes are:
- Function (input/output) driven code for more package-friendly development and not just scripts
- Usage of Config object for more consistent default inputs
- Inputs housed in Github for traceability and flexibility
- Can be run from command line (takes arguments as inputs)
- Runnable on UniCC cluster

## Changes to inputs and parameters
- There now exists a Config object (in `utils.config`) that has the default parameters that should be consistently used in the model, but now we can easily change each parameter via the Config object instead of looking for it in places across different scripts.
- For each "version" of the model (such as "ALPS2023"), make sure a folder with the version name exists in `data/chilled/version` with the following files

## Changes to usage
Instead of modifying scripts that has parameters set in the script, now the parameters can be set via the Config object and modified via arguments. Currently, using the run scripts, these are the arguments that be given from the command line:

- Version name (default: "ALPS2023")
- GCM scenario (default: "GFDL-ESM4")
- RCP (default: "baseline")

Preprocessing of rasters can be done first without considering GCM and RCP. This can be accomplished by running the script (in `/chilled/`):

```
python -m run_preprocess.py -version "version_name"
```

The part of the model that collects and processes the climate data to derive energy intensities can be run using:

```
python -m run_main.py -version "version_name" -gcm "gcm_name" -rcp "rcp_name"
```

Note that the code is now also easily runnable on the UniCC server, which is nice for setting off multiple GCM-RCP runs at once.

## How to review

For @amastrucci or @byersiiasa: please let me know if the code does not make sense now or if the usage is not intuitive/efficient.

Perhaps @glatterf42 can also review if the code structure and docs are okay. 

Thank you!

### CODE REVIEW

File: doc/chilled.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+Input data
+----------
+
+
+Parameters
+----------
+
+In this module, the "version" refers to the version of the model, not the version of the data. For example, the "ALPS2023" version refers to the set of scenario inputs/parameters for the "ALPS2023" run.
+When specifying a version, CHILLED will look in the `/data/chilled/version` directory that the version name exists and that the following files exist as well:
```
Comment (glatterf42) [inline doc/chilled.rst:23]:
```suggestion
When specifying a version, CHILLED will look in the `/data/chilled/version` directory that the version name exists and that the following files exist as well:

```

This does not render as a proper list (which I presume is the intention). I think a blank line before the list should fix this.

To check how it renders, use a venv where you have installed all dependencies you need to render e.g. the message docs. You should then be able to run `sphinx-build -T -E -b html -d _build/doctrees-readthedocs -D language=en . _build/html` inside the `doc` directory and can navigate to the pages when hosting them yourself via `python -m http.server` :)

File: doc/chilled.rst [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+- `arch_input_reg.xlsx` (or `arch_input.xlsx`)
+- `arch_regions.xlsx`
+- `par_var.csv`
+- `runs.csv`
+
+Usage
+=====
+
+The preprocessing (preparing MESSAGE region rasters and country codes) happens without specifying GCM or RCP scenario. Therefore, it can be run first on its own using the following command (from the within :mod:`chilled.run`):
```
Comment (glatterf42) [inline doc/chilled.rst:32]:
```suggestion
The preprocessing (preparing MESSAGE region rasters and country codes) happens without specifying GCM or RCP scenario. Therefore, it can be run first on its own using the following command (from within :mod:`chilled.run`):
```

File: doc/chilled.rst [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+.. code-block:: bash
+
+   python -m main.py -version "version_name" -gcm "gcm_name" -rcp "rcp_name"
+
+If the `-version` command is not provided, then the default version is "ALPS2023". 
+If the `-gcm` and `-rcp` commands are not provided, then the default GCM is "GFDL-ESM2M" and the default RCP is "baseline".
+
+
+.. Code reference
```
Comment (glatterf42) [inline doc/chilled.rst:50]: Is this commented on purpose? 
Is this related to the `:param parsed_arguments:` and `:return:` in the `preprocess/run_*` files? These look weird , too, but maybe this is still work in progress? If not, these docstrings are not rendered anywhere, so don't really provide any information.

File: message_ix_buildings/chilled/main/climate.py
Comment (glatterf42) [inline message_ix_buildings/chilled/main/climate.py:1]: GitHub is not rendering this because the diff is too large, so my review excludes this file.

**Review by glatterf42 — COMMENTED** [review]
Thanks, looks great overall. Some questions and comments in line, though.

File: message_ix_buildings/chilled/functions/user_settings.py
[the remark this answers is not in the document]
Comment (measrainsey) [reply message_ix_buildings/chilled/functions/user_settings.py:1]: Thanks for the catch! I have untracked the file and added it to `.gitignore`.

File: message_ix_buildings/chilled/utils/__init__.py
[the remark this answers is not in the document]
Comment (measrainsey) [reply message_ix_buildings/chilled/utils/__init__.py:1]: Renamed `utils` submodule to `util`.

File: message_ix_buildings/chilled/utils/config.py [hunk 4]
```diff
@@ -81,7 +78,7 @@ class Config:
 
     # climatic inputs for sensitivity runs
     # clims = ["hist"]  # options: "hist", "1p5"
-    clims = list(yeardic.keys())
+    # clims = list(yeardic.keys())
```
[the remark this answers is not in the document]
Comment (measrainsey) [reply message_ix_buildings/chilled/utils/config.py]: Thanks! I went through the config file and tried to remove random commented sections and also cleaned up the file a bit -- hopefully it's better now.

File: message_ix_buildings/chilled/main/__init__.py
[the remark this answers is not in the document]
Comment (measrainsey) [reply message_ix_buildings/chilled/main/__init__.py:1]: Renamed -- from `main` to `core`!

File: doc/chilled.rst [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+:mod:`.chilled` (Cooling and Heating gLobaL Energy Demand) is a model that assesses the impacts of climate change (temperature change)
+
+Methods
+=======
+
+
+Data
+====
+
+Input data
```
[the remark this answers is not in the document]
Comment (measrainsey) [reply doc/chilled.rst:15]: Yes sorry just placeholders for now 🥲 I tried adding some autosummary stuff so at least the functions would appear but my sphinx kept failing to build if I did that.

File: message_ix_buildings/chilled/preprocess/message_raster.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
+"""
+Create raster of MESSAGE regions for CHILLED with specified nodes
+"""
+
+import datetime
+import os
+
+import numpy as np
+import pandas as pd  # type: ignore
```
[the remark this answers is not in the document]
Comment (measrainsey) [reply message_ix_buildings/chilled/preprocess/message_raster.py]: Yeah I was getting linting issues if I didn't have the `type: ignore` but fixed now but installing some stubs for pandas :) Went through and tried to remove the `type: ignore` from pandas and some of the other big packages where I could.

File: doc/chilled.rst (hunk 3 above)
Comment (measrainsey) [reply doc/chilled.rst:50]: Yeah this is commented on purpose, mainly because the automodule and autosummary were not compiling (I encountered errors). I think perhaps I will treat this as a WIP - I will (perhaps in a different PR) create documentation and docstrings for all the different functions and then return to this part of the documentation to make sure the code reference builds properly. Does that sound okay?

Comment (glatterf42) [reply doc/chilled.rst]: Sounds fine to me, but take a look at the latest changes. You could expand that already if you wish :)

**Comment by glatterf42** [comment]
Quickly jumped in here to understand what's going on with the docs: overall, you were moving in the right direction. 

- There were some dependencies missing in `pyproject.toml` (which I noticed since I started in a new venv)
- Sphinx could not understand some of the imports. E.g. importing from `util.config` in `message_ix_buildings/chilled/preprocess/message_raster.py` makes sphinx look at all top-level modules (like `pandas`, `numpy`, or `message_ix_buildings`) and it doesn't find `util` there. So we can either use absolute paths (like I did now) or relative paths. This would then be `from .util.config import Config` in this case. Not entirely sure which one is preferable.

With just these two changes, the docs now built successfully for me, including the code reference part. To replicate (in your venv on a command line of your choice):

1. Go to `doc/`
2. Run a command like `sphinx-build -T -E -b html -d _build/doctrees-readthedocs -D language=en . _build/html`
3. Host your own server via `python -m http.server` (and stop with ctrl+c once you're done)
4. Access the link it shows with your favourite browser (something like `http://0.0.0.0:8000/`)
5. Navigate to `_build/html/` and have a look around :)

**Review by glatterf42 — APPROVED** [review]
This now looks good from my side.

Code could still be cleaned up (e.g. by adding mypy and ruff like we have them in message_ix and addressing linting/formatting errors or by cleaning up comments and `type: ignore` statements), but this could also happen in a later PR. If saved for later, please open an issue to not forget about this. 

Now for @amastrucci, @byersiiasa, or probably @Zh-xy: please review and approve the content-side of the changes :)

**Comment by measrainsey** [comment]
Thanks @glatterf42 for your review and edits! A follow-up:

> * Sphinx could not understand some of the imports. E.g. importing from `util.config` in `message_ix_buildings/chilled/preprocess/message_raster.py` makes sphinx look at all top-level modules (like `pandas`, `numpy`, or `message_ix_buildings`) and it doesn't find `util` there. So we can either use absolute paths (like I did now) or relative paths. This would then be `from .util.config import Config` in this case. Not entirely sure which one is preferable.

Using the absolute paths is giving me an issue now when trying to run the python scripts in command line. For example, with using absolute paths, I am now receiving the error:

```
  File "/home/mengm/repo/message-ix-buildings/message_ix_buildings/chilled/preprocess/message_raster.py", line 12, in <module>
    from message_ix_buildings.chilled.util.config import Config
ModuleNotFoundError: No module named 'message_ix_buildings'
```

Within `message_raster.py`, if I change the import line to use the relative path `from .util.config import Config`, I receive the following error when running the script in command line:

```
  File "/Users/meas/iiasagit/message-ix-buildings/message_ix_buildings/chilled/preprocess/message_raster.py", line 12, in <module>
    from .util.config import Config
ModuleNotFoundError: No module named 'preprocess.util'
```

I figure I needed to move the relative path to the directory before the `preprocess` submodule, but if I tried to change the import line to `from ..util.config import Config` instead, I receive the following error:

```
  File "/Users/meas/iiasagit/message-ix-buildings/message_ix_buildings/chilled/preprocess/message_raster.py", line 12, in <module>
    from ..util.config import Config
ImportError: attempted relative import beyond top-level package
``` 

Do you have any suggestions on how to proceed? The only import line seems to work in order to run the script in command line is `from util.config import Config`, but it seems that's causing issues for the documentation build. 

Thanks!

**Comment by glatterf42** [comment]
Huh, that sounds odd. My first thought was that this might be dependent on where you are when you execute the scripts, but this would rather be an issue with relative imports than with the absolute paths we have now.
My next guess would be that you might not have an editable install of message-ix-buildings, but if you produce various error messages after only changing one line, you likely do have an editable install.
The (maybe unfortunate) thing is: I can't reproduce your error messages. For me, this is what I see:
```bash
fridolin@fridolin-Latitude-5520 ~> (buildings) cd message-ix-buildings/
fridolin@fridolin-Latitude-5520 ~/message-ix-buildings (chilled/restruct)> (buildings) python
Python 3.12.4 (main, Jun  8 2024, 18:29:57) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from message_ix_buildings.chilled.util.config import Config
>>> 
fridolin@fridolin-Latitude-5520 ~/message-ix-buildings (chilled/restruct)> (buildings) cd message_ix_buildings
fridolin@fridolin-Latitude-5520 ~/m/message_ix_buildings (chilled/restruct)> (buildings) python
Python 3.12.4 (main, Jun  8 2024, 18:29:57) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from message_ix_buildings.chilled.util.config import Config
>>> 
fridolin@fridolin-Latitude-5520 ~/m/message_ix_buildings (chilled/restruct)> (buildings) cd chilled
fridolin@fridolin-Latitude-5520 ~/m/m/chilled (chilled/restruct)> (buildings) python
Python 3.12.4 (main, Jun  8 2024, 18:29:57) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from message_ix_buildings.chilled.util.config import Config
>>> 
fridolin@fridolin-Latitude-5520 ~/m/m/chilled (chilled/restruct)> (buildings) python preprocess/message_raster.py 
fridolin@fridolin-Latitude-5520 ~/m/m/chilled (chilled/restruct)> (buildings) git status
On branch chilled/restruct
Your branch is up to date with 'origin/chilled/restruct'.

nothing to commit, working tree clean
fridolin@fridolin-Latitude-5520 ~/m/m/chilled (chilled/restruct)> (buildings) cd ../..
fridolin@fridolin-Latitude-5520 ~/message-ix-buildings (chilled/restruct)> (buildings) python message_ix_buildings/chilled/preprocess/message_raster.py
fridolin@fridolin-Latitude-5520 ~/message-ix-buildings (chilled/restruct)> (buildings)
```
In other words: wherever I am on my system, I can import from message_ix_buildings as from any other package and I can execute `message_raster.py` without error. 

So just to clarify: you pulled the latest changes from here (especially including new `__init__` files) and this directly adapted your install of message-ix-buildings because you installed it via `pip install -e ...`? What are you using to manage your venv? Is the correct venv active? If you run `pip list`, do you see `message-ix-buildings` in that list? With a version like `0.0.2.dev7+gd68f29b.d20240725`? And finally, which OS are you using?

**Comment by byersiiasa** [comment]
Hi all, thanks for this, looks very good.
Main suggestion would be to add a bit more intro to the README to explain more about CHILLED, cite relevant papers, and explain a bit more about the input data types and outputs that are generated.

**Comment by measrainsey** [comment]
> Hi all, thanks for this, looks very good. Main suggestion would be to add a bit more intro to the README to explain more about CHILLED, cite relevant papers, and explain a bit more about the input data types and outputs that are generated.

Thanks Ed! If it's alright, my preference at the moment would be to get this PR merged and then add the documentation in another PR (so that this one does not get too large). 🙂 I have created an issue for the documentation: #21

**Comment by measrainsey** [comment]
> Code could still be cleaned up (e.g. by adding mypy and ruff like we have them in message_ix and addressing linting/formatting errors or by cleaning up comments and `type: ignore` statements), but this could also happen in a later PR. If saved for later, please open an issue to not forget about this.

As a note, I just created an issue for this: #22

**Comment by byersiiasa** [comment]
Its only for Fridolin to approve ;)

**Comment by measrainsey** [comment]
@glatterf42 If it's okay could I go ahead and merge? :)

File: message_ix_buildings/chilled/preprocess/archetypes.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    # get raster file and message map
+    country_ras, reg_ras, map_reg, iso_attrs = create_message_raster(config)
+
+    # get archs
+    vers_archs = get_archs(config)
+
+    # save MESSAGE regions map
+    msg_file = "map_reg_MESSAGE_" + config.node + ".nc"
+    map_reg.to_netcdf(os.path.join(archetype_path, msg_file))
+    print(
```
Comment (glatterf42) [inline message_ix_buildings/chilled/preprocess/archetypes.py:39]: Just a note for the future: such information is best printed via [python's logging utility](https://docs.python.org/3/howto/logging.html#logging-basic-tutorial). Please consider improving that in the future (and you could track this with an issue, too, if you'd like).

File: message_ix_buildings/chilled/preprocess/message_raster.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            "authors": "Edward Byers & Alessio Mastrucci",
+            "date": str(datetime.datetime.now()),
+            "institution": "IIASA Energy Program",
+            "contact": "byers@iiasa.ac.at; mastrucc@iiasa.ac.at",
+        }
+
+        return country_ras, reg_ras, map_reg, iso_attrs
+
+    else:
+        print("Only R11 is supported at the moment.")
```
Comment (glatterf42) [inline message_ix_buildings/chilled/preprocess/message_raster.py:127]: Should this really just `print()` or raise an error (preventing the code from continuing to run)?

File: message_ix_buildings/chilled/run_agg.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    # process_construction_shares(cfg)
+
+    # print("RUNNING process_floor_area_maps()........")
+    # process_floor_area_maps(cfg)
+
+    # print("RUNNING process_country_maps()........")
+    # process_country_maps(cfg)
+
+    # print("RUNNING process_final_maps()........")
+    # process_final_maps(cfg)
```
Comment (glatterf42) [inline message_ix_buildings/chilled/run_agg.py:91]: All these comments might warrant another line of explanation: "to achieve x, uncomment y" or so. Your future you will likely thank you :)

File: message_ix_buildings/chilled/run_main.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        (process_floor_area_maps(step),)
+        (process_country_maps(step),)
+        (process_final_maps(step),)
+        (process_iso_tables(step),)
+
+
+if __name__ == "__main__":
+    main()
+if __name__ == "__main__":
+    main()
```
Comment (glatterf42) [inline message_ix_buildings/chilled/run_main.py:101]: Why is this function/run condition here twice?

File: message_ix_buildings/chilled/run_preprocess.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+#     os.path.join(cfg.dle_path, "out", "raster", "map_reg_MESSAGE_" + cfg.node + ".nc")
+# ):
+#     create_message_raster(cfg)
+# else:
+#     print(f"map_reg_MESSAGE_{cfg.node}.nc already exists. Using existing file.")
+
+# run create_archetypes
+
+
+# create maps of archetype variables
```
Comment (glatterf42) [inline message_ix_buildings/chilled/run_preprocess.py:82]: Are these comments here on purpose? If so, what is it?

File: message_ix_buildings/chilled/util/config.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    #: List all possible RCP scenarios.
+    rcps = ["ssp126", "ssp370", "ssp585", "baseline"]
+
+    #: Set what data to use for the RCP scenario.
+    #: If rcp is "baseline", then use "ssp126" data.
+    #: Otherwise, use the data corresponding to the RCP scenario selected.
+    if rcp == "baseline":
+        rcpdata = "ssp126"
+    else:
+        rcpdata = rcp
```
Comment (glatterf42) [inline message_ix_buildings/chilled/util/config.py:45]:
```suggestion
    rcpdata = "ssp126" if rcp == "baseline" else rcp
```
Something like this should also work and is generally regarded "less complex".

**Review by glatterf42 — APPROVED** [review]
Almost, yeah, that's why I approved this PR before :)

Since you asked for a new review, I've looked again at the files your recent commits changed and found some hints and questions that I hope are useful. I trust you will handle them well, so I will approve this PR again and once you're satisfied with your actions responding to the new comments, please feel free to merge :)

File: message_ix_buildings/chilled/util/config.py (hunk 12 above)
Comment (glatterf42) [reply message_ix_buildings/chilled/util/config.py:45]: (Might be that the line is then too long for ruff's liking or so.)

File: message_ix_buildings/chilled/preprocess/archetypes.py (hunk 7 above)
Comment (measrainsey) [reply message_ix_buildings/chilled/preprocess/archetypes.py:39]: Thanks! That's a very good point -- I have created issue #24 to track this.

File: message_ix_buildings/chilled/run_preprocess.py (hunk 11 above)
Comment (measrainsey) [reply message_ix_buildings/chilled/run_preprocess.py]: Not sure why they're there, thanks for the catch! Removed :)

File: message_ix_buildings/chilled/run_agg.py (hunk 9 above)
Comment (measrainsey) [reply message_ix_buildings/chilled/run_agg.py]: Think I commented these out when I was debugging, but they should all be uncommented, so I did that for now.

File: message_ix_buildings/chilled/preprocess/message_raster.py (hunk 8 above)
Comment (measrainsey) [reply message_ix_buildings/chilled/preprocess/message_raster.py]: Good call! Changed from `print()` to a `raise TypeError("...")`
